// @thoughtspot/visual-embed-sdk/cjs/src/utils/graphql/nlsService/nls-answer-service.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("jest-fetch-mock");
const nls_answer_service_1 = require("./nls-answer-service");
describe('getAnswerFromQuery', () => {
    test('should return answer and suggestion', async () => {
        fetchMock.mockResponses(JSON.stringify({
            data: {
                queryRequest: {
                    sageQuerySuggestions: [
                        {
                            sessionId: 'sessionId',
                            genNo: 2,
                            stateKey: {
                                transactionId: 'transactionId',
                                generationNumber: 1,
                            },
                        },
                    ],
                },
            },
        }));
        const query = 'query';
        const worksheetId = 'worksheetId';
        const res = await (0, nls_answer_service_1.getAnswerFromQuery)(query, worksheetId);
        const session = res.answer.getSession();
        expect(session.sessionId).toEqual('sessionId');
        expect(session.genNo).toEqual(2);
        expect(session.acSession.sessionId).toEqual('transactionId');
        expect(session.acSession.genNo).toEqual(1);
    });
});
